### DISCUSSION TITLE
make SBOM without composer-plugin

### DISCUSSION THREAD

**Discussion by jkowalleck** [CycloneDX/cyclonedx-php-composer#103] [Ideas]
current implementation is a composer plugin.
this has some flaws: 

* cannot generate the SBOM for lock files outside of the current composerized project.
  can generate SBOM only based in current project.  
  
  this is not an issue at the moment, since this plugin does not utilize any composer-requirements on its own.   
  so current switches `--exclude-dev` or `--exclude-plugins` can exclude _cyclonedx-php-composer_ at the moment.  
  
  in the future it is planned to change, to incorporate [`package-url/packageurl-php`](https://packagist.org/packages/package-url/packageurl-php).
  with this change, the plugin will ship an own dep, which is not excludable via the `--exclude-dev` switch at the moment.

----

IDEA: ship a "binary" that can generate SBOM based on a lock file, without leverage of composer-plugin

BENEFIT: 
* generated SBOM can be built without requirements of _cyclonedx-php-composer_.

----

THOUGHTS:
* its planned to have a `cyclonedx-php-core` which holds all needed models, functionality ...
* why not ship a php script as "binary" with it, that can generate SBOM based on any lock file?

**Comment by jkowalleck** [comment]
possible implementation: build a executable [PHAR](https://www.php.net/manual/en/intro.phar.php)
that bundles composer, this very composer-plugin and all dependencies.

this way there was a shippable binary that contained all features.

**Reply by theofidry** [reply]
My two cents here hoping this describe the need clearly. I think you should have a core library which has something along the lines of:

```php
SbomManifestFactory::fromComposerLock(string $composerLockContents): string
```

This would have several benefits:
- Building an app around it to provide a convenient CLI application, is a lot easier. You could use SymfonyConsole, LaravelConsole or whatever it would like be just a thin wrapper. Creating a PHAR to distribute a binary would then be easy too.
- For tools like Box/BoxManifest or other, this would allow to generate the manifest without requiring to launch a console command.

**Reply by jkowalleck** [reply]
currently not intended. 
This project is about composer projects, and therefore it will not use arbitrary `string` nor other data formats, it will use `composer` native data models like  [`RootPackageRepository`](https://github.com/composer/composer/blob/main/src/Composer/Repository/RootPackageRepository.php) and 
 [`Locker`](https://github.com/composer/composer/blob/main/src/Composer/Package/Locker.php)


> For tools like Box/BoxManifest or other, this would allow to generate the manifest without requiring to launch a console command.

**Comment by jkowalleck** [comment]
idea closed.
idea is outdated. all the things discussed were implemented already, the project is able to anayze projects outside of its own context, now.
